### PR TITLE
Add in consul datacenter (AvailabilityZone) support

### PR DESCRIPTION
**Pull request by liamawhite**
What this PR does / why we need it: Currently we only support AZ routing for kube based service registries. This enables it for consul.

Which issue this PR fixes:
Continues on from: istio/old_pilot_repo#1230, #2054 and #1815
Working towards: #1473

Release note:

Enable consul intra-datacenter routing with AvailabilityZone

### CODE REVIEW

**Comment by rshriram** [comment]
I remember seeing dc1 being hardcoded in the init code for Consul (pilot-discovery/...)..

**Comment by GregHanson** [comment]
pilot-discovery code shriram mentioned: https://github.com/istio/istio/blob/master/pilot/cmd/pilot-discovery/server/server.go#L434

**Comment by liamawhite** [comment]
@GregHanson @rshriram so how should we be retrieving the datacenter. You can only get the dc by retrieving a service in the consul API from what I can see. Is that right?

**Comment by rshriram** [comment]
@anubhavmishra any thoughts on this?

**Comment by anubhavmishra** [comment]
@rshriram I am not sure about the context when it comes to what AZ support means for Consul. But here is some insights into Consul API with respect to the datacenter information.
* To get the datacenter information for the agent running on a node you can query `/agent/self`: https://www.consul.io/api/agent.html#read-configuration
* To get all known data centers, you can use the Consul's catalog endpoint `/catalog/datacenters`: https://www.consul.io/api/catalog.html#list-datacenters

I hope this helps. I would love to get more context on what we are trying to achieve here :)

**Comment by rshriram** [comment]
sorry for the lack of context.. We are trying to enable availability zone aware load balancing in Envoy. To do so, we need to provide envoy with the availability zone of the node its running on, so that it will attempt to load balance requests to other services in teh same AZ. (In consul terminology, DC is same as the AZ being discussed here). 

Hope that helps.

**Comment by anubhavmishra** [comment]
I that case, A consul datacenter can span multiple AZs. So not sure if I would use datacenter as the zone identifier.

**Comment by liamawhite** [comment]
@anubhavmishra So the idea of this feature is to keep requests within the same datacenter (mostly for latency reason) by preferring to route to upstream services within the same AZ if possible. Is there anything that identifies a specific geographic availability zone?

**Comment by anubhavmishra** [comment]
@liamawhite yea, so DC doesn't mean an AZ but you can totally create a datacenter that has nodes in only one AZ. In Consul itself there is no identifier to figure out what AZ a node is running in, it is not seen as Consul's responsibility to figure out cloud provider level implementation details. You can totally tag Consul services in the catalog to have this data. 

Another thing you can take a look at is Network Tomography data that Consul provides in a first class way. Here are the docs: https://www.consul.io/docs/internals/coordinates.html

TLDR; you can figure out the roundtrip time information using Consul and do load balancing that way but I am sure this will require a lot of work to implement in Istio.

**Comment by liamawhite** [comment]
@rshriram @GregHanson ready for review
